From 2ce0fc27d2c8814e53d9e02a1f8c65a0f68f04d9 Mon Sep 17 00:00:00 2001
From: Balamurugan S <quic_bselvara@quicinc.com>
Date: Sat, 28 Jan 2023 11:01:24 +0530
Subject: [PATCH 2/2] ath12k: Introduce module param for NAPI budget processing

Increase NAPI poll budget to 0x7f. In rx frames processing by default
64 frames are processed per interrupt. This is increased to 127.

Signed-off-by: Balamurugan S <quic_bselvara@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pci.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -52,6 +52,11 @@ EXPORT_SYMBOL(ath12k_fw_mem_seg);
 module_param_named(fw_mem_seg, ath12k_fw_mem_seg, uint, 0644);
 MODULE_PARM_DESC(fw_mem_seg, "Enable/Disable FW segmentted memory");
 
+unsigned int ath12k_napi_poll_budget = 0x7f;
+EXPORT_SYMBOL(ath12k_napi_poll_budget);
+module_param_named(napi_budget, ath12k_napi_poll_budget, uint, 0644);
+MODULE_PARM_DESC(napi_budget, "Napi budget processing per rx intr");
+
 static const struct pci_device_id ath12k_pci_id_table[] = {
 	{ PCI_VDEVICE(QCOM, QCN9274_DEVICE_ID) },
 	{ PCI_VDEVICE(QCOM, WCN7850_DEVICE_ID) },
@@ -492,7 +497,7 @@ static irqreturn_t ath12k_pci_ext_interr
 static int ath12k_pci_ext_irq_config(struct ath12k_base *ab)
 {
 	int i, j, ret, num_vectors = 0;
-	u32 user_base_data = 0, base_vector = 0, base_idx;
+	u32 user_base_data = 0, base_vector = 0, base_idx, budget;
 	struct ath12k_pci *ar_pci = (struct ath12k_pci *)ab->drv_priv;
 
 	base_idx = ATH12K_PCI_IRQ_CE0_OFFSET + CE_COUNT_MAX;
@@ -503,6 +508,9 @@ static int ath12k_pci_ext_irq_config(str
 	if (ret < 0)
 		return ret;
 
+	if (ath12k_napi_poll_budget < NAPI_POLL_WEIGHT)
+		ath12k_napi_poll_budget = NAPI_POLL_WEIGHT;
+
 	for (i = 0; i < ATH12K_EXT_IRQ_GRP_NUM_MAX; i++) {
 		struct ath12k_ext_irq_grp *irq_grp = &ab->ext_irq_grp[i];
 		u32 num_irq = 0;
@@ -510,8 +518,14 @@ static int ath12k_pci_ext_irq_config(str
 		irq_grp->ab = ab;
 		irq_grp->grp_id = i;
 		init_dummy_netdev(&irq_grp->napi_ndev);
+
+		if (ab->hw_params->ring_mask->rx_mon_dest[i])
+			budget = NAPI_POLL_WEIGHT;
+		else
+			budget = ath12k_napi_poll_budget;
+
 		netif_napi_add(&irq_grp->napi_ndev, &irq_grp->napi,
-			       ath12k_pci_ext_grp_napi_poll, NAPI_POLL_WEIGHT);
+			       ath12k_pci_ext_grp_napi_poll, budget);
 
 		if (ab->hw_params->ring_mask->tx[i] ||
 		    ab->hw_params->ring_mask->rx[i] ||
